zio: Rework Attributes to be Descriptions

teburd opened the pull request:
Rather than providing the attribute values directly in get_attrs
for zio devices and channels, provide descriptions of each attribute.

This should allow a driver to map many attributes directly to registers
or internalized struct members and avoid storing these values in a
zio_variant which is confusing.

Signed-off-by: Tom Burdick <thomas.burdick@gmail.com>

microbuilder commented:
@bfrog Just looking through this, but it might make sense to have two commits to review, one for the core zio API changes and the second for the synth driver updates. The driver is small enough it's not hard to review together, but maybe worth doing like that for other reviewers in the future. I do agree attributes seem a bit cumbersome as-is, though, at least on reasonably complex devices.

microbuilder commented on include/zio/zio_dev.h:
```diff
@@ -141,6 +147,37 @@ struct zio_dev_chan {
+ * Example: #define MY_ATTR_TYPE (ZIO_ATTR_TYPES+1)
+ */
+enum zio_attr_type {
+	/* samples per second */
```
Enum records generally have the comment **after** the field, with `/**<` to indicate that the comment block refers to the preceding field, so:

```
ZIO_SAMPLE_RATE,
/**< @brief samples per second */
...
```

microbuilder commented on include/zio/zio_dev.h:
```diff
@@ -141,6 +147,37 @@ struct zio_dev_chan {
+/** Attribute record. */
+struct zio_attr_desc {
+	/* Attribute type, describing what it is such as the sample rate */
+	enum zio_attr_type type;
```
Doxygen comment blocks need to start with `**`, and would benefit from the `@brief` tag before the text.

Struct attributes generally have the comment **after** the field, though, with `/**<` to indicate that the comment block refers to the preceding field, so:

```
enum zio_attr_type type;
/**< @brief Attribute type, describing what it is such as the sample rate */

...
```

microbuilder reviewed (changes requested):
Looks good other than some doxygen issues commented inline.

dbkinder commented on include/zio/zio_dev.h:
```diff
@@ -80,14 +79,21 @@ struct zio_buf;
+ * Extended values should always be based on ZIO_CHAN_TYPES count with an
  * additive and defined in a publicly available driver header.
+ *
+ * Example: #define MY_CHAN_TYPE (ZIO_CHAN_TYPES+1)
```
``#name`` in a doxygen comment is used to reference a variable named "name", hence the error that's showing up saying:
```
warning: explicit link request to 'define' could not be resolved
```
Workaround is to escape the ``#`` with ``\#``

```suggestion
 * Example: \#define MY_CHAN_TYPE (ZIO_CHAN_TYPES+1)
```

dbkinder commented on include/zio/zio_dev.h:
```diff
@@ -129,18 +135,52 @@ struct zio_dev_chan {
+ * Extended values should always be based on ZIO_ATTR_TYPES count with an
+ * additive and defined in a publicly available driver header.
+ *
+ * Example: #define MY_ATTR_TYPE (ZIO_ATTR_TYPES+1)
```
```suggestion
 * Example: \#define MY_ATTR_TYPE (ZIO_ATTR_TYPES+1)
```

dbkinder reviewed (changes requested):
fix doc warnings from doxygen comments

teburd committed:
zio: attribute descriptions

Rather than providing the attribute values directly in get_attrs
for zio devices and channels, provide descriptions of each attribute.

This should allow a driver to map many attributes directly to registers
or internalized struct members and avoid storing these values in a
zio_variant which is confusing.

Signed-off-by: Tom Burdick <thomas.burdick@gmail.com>

teburd committed:
zio: synth: use attribute descriptions

Update synth driver and sample to use attribute descriptions

Signed-off-by: Tom Burdick <thomas.burdick@gmail.com>

teburd committed:
zio: documentation

Add placeholder and fix duplicated label to in docs
to make ci happy

Signed-off-by: Tom Burdick <thomas.burdick@gmail.com>

bfrog commented:
I believe I've address all the concerns @dbkinder @microbuilder

dbkinder commented on doc/reference/zio/index.rst:
```diff
@@ -0,0 +1,20 @@
+ZIO Attribute Definitions
+*************************
+
+.. doxygengroup:: zio_attributes
```
This doxygen group isn't defined (in this PR)

bfrog replied:
doh, thanks, will fix this up

bfrog replied:
I've removed that doxygen group for now

dbkinder reviewed (changes requested):
When I pull your PR locally, I see a file ``samples/zio/synth/README.rst that doesn't appear to actually part of your PR file list.  Not sure what's up (maybe a rebase is needed)?

bfrog commented:
@dbkinder I'm still not sure what the documentation problems here are exactly, any clues would be helpful

MaureenHelm reviewed (approved):
@dbkinder please revisit

dbkinder reviewed (approved):
I'm not seeing any doc build errors when I generate docs locally, so LGTM.

MaureenHelm commented:
> I'm not seeing any doc build errors when I generate docs locally, so LGTM.

Thanks!
